python/ray/util/client/dataclient.py

```diff
@@ -59,48 +59,122 @@ def _start_datathread(self) -> threading.Thread:
         return threading.Thread(target=self._data_main, args=(), daemon=True)
 
     def _data_main(self) -> None:
-        stub = ray_client_pb2_grpc.RayletDataStreamerStub(self.channel)
-        resp_stream = stub.Datapath(
-            iter(self.request_queue.get, None),
-            metadata=self._metadata,
-            wait_for_ready=True)
+        reconnecting = False
         try:
-            for response in resp_stream:
-                if response.req_id == 0:
-                    # This is not being waited for.
-                    logger.debug(f"Got unawaited response {response}")
-                    continue
-                if response.req_id in self.asyncio_waiting_data:
-                    callback = self.asyncio_waiting_data.pop(response.req_id)
-                    try:
-                        callback(response)
-                    except Exception:
-                        logger.exception("Callback error:")
-                else:
-                    with self.cv:
-                        self.ready_data[response.req_id] = response
-                        self.cv.notify_all()
-        except grpc.RpcError as e:
+            while not self.client_worker._in_shutdown:
```

Comment:
I refactored this logic in #18539 to use a non-reentrant lock to protect all mutable states of DataClient. I think it made the logic easier to reason about. Right now, even if everything could be correct, no lost messages during shutdown etc, it can still be a bit hard to reason about.

Reply:
Sounds good, 18539 has been merged into master + this PR